Describe DpMechanism interface - how noise mechanisms should generate noise

junyechen1996 opened the pull request:
- Note to reviewers: this is stacked PR on #2 .

wangshan commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -255,12 +255,58 @@ with the correct noise type.
 client measurement and/or an aggregation.
 
-> TODO: Junye to spell out the interface of dp mechanisms, "Noise Mechanisms to Accomplish DP"
+Clients and Aggregators can implement various noise mechanisms to accomplish
```
Noise mechanisms and DP mechanisms are mixed, we should stick to one of them, I’d vote for DP mechanisms

junyechen1996 replied:
Changed all to DP mechanisms

wangshan commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -255,12 +255,58 @@ with the correct noise type.
+        """
+        Debias the data due to the added noise.
+        This doesn't apply to all noises. Some local randomizers
+        need this functionality.
```
Avoid mentioning local randomizer unless we are gonna explain it

junyechen1996 replied:
Removed local randomizer, instead call it Client-DP.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -255,12 +255,58 @@ with the correct noise type.
+        """
+        pass
+
+    def debias(self,
```
How about 
```suggestion
    def remove_noise_from_result(self,
```

junyechen1996 replied:
The set of local DP noises we have in mind don't remove noise completely, e.g. see Algorithm 2 and Algorithm 4 in [{{MJTBp22}}](https://arxiv.org/abs/2004.00010), the aggregate result will still have noise, with some standard deviation. I think `debias` is a better term because it is "removing" the positive bias added in the local DP.

cjpatton replied:
Ahh right, "remove noise" is not a good term. How about "remove bias"? The problem I'm trying to solve is that "debias" doesn't look like "de-bias" to me, it looks like "deebee-ass" :D

wangshan replied:
Debias has been used in referenced literatures so it should be clear to readers, remove_bias would do too but I don’t know if we can avoid mentioning debias

junyechen1996 replied:
Sticking with `debias` for now. We can always change it to `remove_bias` if others feel it is better as well.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -255,12 +255,58 @@ with the correct noise type.
+        """Add noise to a piece of data. """
+        pass
+
+    def sample_noise(self, num_reps: int, dimension: int) -> DataType:
```
IIUC this would be called by add_noise, i.e., is more of a DP primitive than a DP mechanism. Can we rename it?

junyechen1996 replied:
I think this gives caller flexibility to encode the returned noise differently, e.g. when sampling noise in the VDAF setting, one will sample the noise with this method, and then encode it to into field elements. To take DiscreteGaussian as an example, the returned data is an array of signed integers, and there is another step that converts the negative values to the back end of field modulus (e.g. -1 to `q-1`, where `q` is field modulus).

cjpatton replied:
Think of discrete guassian as a primitive that we use to implement `add_noise_to_agg_share()` or `add_noise_to_measurement()`. We would invoke that primitive in both of these methods, then do any postprocessing that's needed to fit with the VDAF. Does that make sense?

junyechen1996 replied:
Sounds good, moved those logic into `DpMechanism`. `DpPrimitive`'s `sample_noise` will only deal with generating noise.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -255,12 +255,58 @@ with the correct noise type.
+    DataType
+    DebiasedDataType
+
+    def add_noise(self, data: DataType) -> DataType:
```
Add noise into what, the measurement or the aggregate share? How do we know when to call this?

I think it would be worth splitting this class into `ClientDpMechanism` and `AggregatorDpMechanism`.

junyechen1996 replied:
Currently I was going for both, adding noise into measurement or aggregate share, because the noise itself doesn't know about measurement or aggregate share. Assume we turn this PR into tackling primitives instead after #9 is merged, a noise primitive should just have the flexibility of adding noise to data directly, or sampling noise that allows caller to perform future operations on it.

cjpatton replied:
> Currently I was going for both, adding noise into measurement or aggregate share, because the noise itself doesn't know about measurement or aggregate share.

I see. In my opinion these should be split into two different classes, as this results in the simplest, most misuse-resistant API surface. Consider an implementation that incorrectly uses a `DpMechanism` intended for Client-DP as a mechanism for Aggregator-DP: Could this result in the aggregate result not being DP, for some definition of DP? In particular, don't we have to tune the sensitivity differently in each case?

> Assume we turn this PR into tackling primitives instead after #9 is merged, a noise primitive should just have the flexibility of adding noise to data directly, or sampling noise that allows caller to perform future operations on it.

Remember, by "primitive" I mean low-level algorithms used to build DP mechanisms. I think this PR should focus on the interface for the latter (DP mechanisms).

wangshan replied:
The idea is adding noise to measurement and adding noise to aggregate share are fundamentally the same operation, having two classes is an implementation detail. We can also have generic add_noise function that applies to Measurement and AggShare types.

cjpatton replied:
The add_noise function that is common would be specified in {{primitives}}; in this section, we would say how the primitive is used in a particular Client-DP mechanism or Aggregator-DP mechanism.

cjpatton replied:
@junyechen1996 and I discussed offline and converged on the idea of a single, unified `DpMechanism` interface. It turns out that this is really the only viable choice, as there are mechanisms that involve both client and aggregator adding noise, but for which the client bits and aggreagtor bits aren't secure on their own. Thus they MUST be unified.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -255,12 +255,58 @@ with the correct noise type.
+
+* `sample_noise` that allows `DpMechanism` to sample noise directly. This can
+  be used by an Aggregator to sample Aggregator DP noise, which then gets merged
+  with the Aggregator's aggregate share.
```
Is this needed at this level of abstraction? IIUC sample_noise would be called by an implementation of add_noise and so seems like an implementation detail here.

junyechen1996 replied:
I am trying to explain why `sample_noise` exists, related to the reason that I explained here: https://github.com/wangshan/draft-wang-ppm-differential-privacy/pull/4#discussion_r1267188326

junyechen1996 commented:
Reviewers, I try to take into account of suggestions from both of you, following changes are proposed:

- I moved the previous `DpMechanism` interface into `DpPrimitive`, I think those interfaces make sense for DP primitives, i.e. they don't know about measurement vs aggregate share. I have some initial texts for DP primitives and their functionalities, but haven't described a concrete example.
- I proposed `ClientDpMechanism` and `AggregatorDpMechanism`. I think `DpPrimitive.add_noise` itself won't always work for both measurement and aggregate share, which could be of different types. It's also more error proof when one tries to instantiate it, related to the discussion in [this comment](https://github.com/wangshan/draft-wang-ppm-differential-privacy/pull/4#discussion_r1266142207).
- I described the RR Client-DP and Aggregator-DP mechanisms in details. Let me know if it's informative.

I force pushed by accident during rebasing, but I think the new version of PR is quite different from the previous one. Let me know of additional feedbacks.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,9 +403,55 @@ with the correct noise type.
+
+We also expect DP primitives to contain the following functionalities:
+
+* Add noise to a piece of input data. Some DP primitives apply noise based on
```
Matching terminology in the DAP/VDAF specs helps avoid ambiguity.

```suggestion
* Add noise to a piece of input data (i.e., a measurement or aggregate share). Some DP primitives apply noise based on
```

junyechen1996 replied:
Missed this one, but done.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,9 +403,55 @@ with the correct noise type.
+  primitive that flips the bit at each dimension, which means the noised output
+  depends on the input.
+
+* Sample noise directly.
```
```suggestion
* Sample noise using one of the primitives in {{primitives}}.
```

junyechen1996 replied:
Done.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,9 +403,55 @@ with the correct noise type.
+We also expect DP primitives to contain the following functionalities:
+
+* Add noise to a piece of input data. Some DP primitives apply noise based on
+  the input data, e.g. Appendix C of {{MJTBp22}} describes a randomized response
```
We should refer to our own spec instead of the appendix of a paper.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,9 +403,55 @@ with the correct noise type.
+* Sample noise directly.
+
+* Debias the noised data. Note that not all noise will need this functionality.
+  We think some Client-DP noises will need this functionality. For example,
```
"Client-DP noises"? Do you mean mechanism?

junyechen1996 replied:
Done.

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+    def maybe_add_noise_to_agg_share(self,
+                                     agg_share: AggregateShare,
+                                     meas_count: Unsigned,
+                                     min_batch_size: Unsigned,
```
@cjpatton I changed it back to `min_batch_size`, because I think `batch_size` might mean the actual batch size.

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+        """
+        if meas_count >= min_batch_size:
+            return self.rr_primitive.debias(agg_result, meas_count)
+        # The count passed to debiasing will be:
```
This is the tricky part about debiasing, when there is composition of Client-DP and Aggregator-DP.

cjpatton replied:
Let's use the terms "client noise" and "aggregator noise"; we don't want to conflate the "process" (adding noise) with the "result" (DP)

junyechen1996 replied:
I didn't explicitly explain Client-DP and Aggregator-DP noise in this function's comments. Is there anything else that I missed?

cjpatton replied:
LGTM

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+    DebiasedAggregateResult = List[int]
+
+    def __init__(self,
+                 dgauss_primitive: DpPrimitive,
```
We haven't fully specified discrete Gaussian primitive, so leaving it as an initialized parameter. But once we have something for discrete Gaussian, we can update both here and for randomized response.

cjpatton replied:
SGTM. Just put a TODO here, and maybe link to the #10 .

junyechen1996 replied:
Done.

cjpatton commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+class DpMechanism:
+    Measurement
+    AggregateShare
+    AggregateResult
+    DebiasedAggregateResult
```
nit: I think it's safe to assume that AggregateResult and DebiasedAggregateResult are always the same.

wangshan replied:
One difference is when the aggregate result is [integer] and DebiasedAggregateResult is [float]

cjpatton replied:
Oh, OK! Couldn't the aggreagte result also be a float?

junyechen1996 replied:
Currently `Prio3FixedPointBoundedL2VecSum` uses aggregate result as an array of float, so that could be the case. And there is nothing needed in debiasing.

cjpatton replied:
I suggest having one type for now until we need the flexibility.

junyechen1996 replied:
I think `AggregateResult` is a term defined in VDAF draft. For `Prio3SumVec`, it is an array of unsigned integers. If RR is used, Collector needs to debias aggregate result, which will turn it from an array of unsigned integers to an array of floats. It seems like the distinction is needed here?

cjpatton replied:
Why should that be the case? If what we want is vector of sums, then we'd expect integers wouldn't we? Why convert them to floats?

junyechen1996 replied:
Because the arithmetic operations in debiasing results in floats, see Appendix A.2 in https://arxiv.org/pdf/2211.10082.pdf, and also could result in negative values. So for Prio3SumVec for example, aggregate result is an array of *unsigned* integers, debiased aggregate result needs to be at least an array of *signed* integers, or floats.

cjpatton replied:
Ack. OK, let's kep it then

wangshan commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+    AggregateResult
+    DebiasedAggregateResult
+
+    def maybe_add_noise_to_measurement(self,
```
I don’t think we should have maybe in the interface, the fact it’s optional is something the implementation language should handle

cjpatton replied:
SGTM.

cjpatton replied:
(Adjusted in slides.)

junyechen1996 replied:
Updated PR as well to `add_client_dp`, `add_aggregator_dp`, and `debias_aggregate_result`.

wangshan commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+    Measurement = int
+    AggregateShare = List[Field]
+    AggregateResult = List[int]
+    DebiasedAggregateResult = List[int]
```
DebiasedAggregateResult = AggregateResult
to make it more clean

junyechen1996 replied:
Done.

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -400,37 +399,270 @@ with the correct noise type.
+        """
+        pass
+
+    def debias_aggregate_result(self,
```
@cjpatton Is it worth being explicit here (`debias_aggregate_result` vs `debias`), if the argument takes in an aggregate result? If you think it's good, then the slides need update. Otherwise, I can revert it.

cjpatton replied:
Good catch, yes debias_aggregate_result is better IMO. Fixing the slides.

junyechen1996 committed:
Texts for DP mechanisms and policies

- Define interface for DP mechanisms
- Define interface for DP policies, actions for Client, Aggregator, and Collector
- Describe concrete DP policies for Prio3SumVec with RR.
- Describe concrete DP policies for Prio3Histogram with discrete Gaussian.

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,31 +403,266 @@ This section describes various mechanisms required for implementing DP
+                          agg_result: AggregateResult,
+                          meas_count: Unsigned,
+                          min_batch_size: Unsigned,
+                          num_aggregators: Unsigned,
```
@cjpatton Added `num_aggregators` parameter in this method, which is needed by `Prio3SumVecWithRandomizedResponse`.

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,31 +403,266 @@ This section describes various mechanisms required for implementing DP
+
+~~~
+class Prio3SumVecWithRandomizedResponse:
+    Measurement = Vec[Unsigned]
```
Using `Vec` and `Unsigned` to be consistent with VDAF draft.

junyechen1996 commented on draft-wang-ppm-differential-privacy.md:
```diff
@@ -403,31 +403,266 @@ This section describes various mechanisms required for implementing DP
+class Prio3HistogramWithDiscreteGaussian:
+    Measurement = Unsigned
+    AggregateShare = Vec[Field]
+    AggregateResult = Vec[int]
```
There is some small caveat about the aggregate result type for `Prio3Histogram` with discrete Gaussian, because discrete Gaussian can sample negative noise, so it's possible that aggregate result becomes negative. In the Field world, it is at the tail end of field modulus.

cjpatton reviewed (approved):
I didn't manage a thorough review, but I think this is definitely good enough for now! 🚢

junyechen1996 commented:
squashed, going to merge.
